tls/s2n_server_hello.c

```diff
@@ -36,6 +37,54 @@
 /* From RFC5246 7.4.1.2. */
 #define S2N_TLS_COMPRESSION_METHOD_NULL 0
 
+/* From RFC8446 4.1.3. */
+#define S2N_DOWNGRADE_PROTECTION_SIZE   8
+const uint8_t tls12_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x01
+};
+
+const uint8_t tls11_downgrade_protection_bytes[] = {
+    0x44, 0x4F, 0x57, 0x4E, 0x47, 0x52, 0x44, 0x00
+};
+
+static int s2n_detect_downgrade(struct s2n_connection *conn) {
+    notnull_check(conn);
+    uint8_t *downgrade_protection = &conn->secure.server_random[S2N_TLS_RANDOM_DATA_LEN - S2N_DOWNGRADE_PROTECTION_SIZE];
+
+    /* Detect downgrade attacks according to RFC 8446 section 4.1.3 */
+    if (s2n_is_tls13_enabled() && conn->server_protocol_version == S2N_TLS12) {
+        /* If the server supports TLS1.3, the random data will be set according to RFC sect 4.1.3 */
+        if (memcmp(downgrade_protection, tls12_downgrade_protection_bytes, S2N_DOWNGRADE_PROTECTION_SIZE) == 0) {
+            GUARD(s2n_queue_reader_illegal_parameter_alert(conn));
+            S2N_ERROR(S2N_ERR_SERVER_PROTOCOL_VERSION);
+        }
+    } else if (conn->server_protocol_version < S2N_TLS12) {
+        /* If the server supports TLS1.2, the random data should be set according to RFC sect 4.1.3 */
+        if (memcmp(downgrade_protection, tls11_downgrade_protection_bytes, S2N_DOWNGRADE_PROTECTION_SIZE) == 0) {
+            GUARD(s2n_queue_reader_illegal_parameter_alert(conn));
+            S2N_ERROR(S2N_ERR_SERVER_PROTOCOL_VERSION);
+        }
+    }
+
+    return 0;
+}
+
+static int s2n_add_downgrade_mechanism(struct s2n_connection *conn) {
```

Comment:
should this always be applied or only when the server version is running in TLS13?

Reply:
It should always be applied. The only difference is what the downgraded protocol version is.

Reply:
this function should be guarded with s2n_is_tls13_enabled() as we do not want to apply this mechanisms if we are not running a "tls 13 server".

Reply:
The RFC says TLS1.2 server SHOULD apply downgrade protection, so I didn't want to make it tls13 only.

Reply:
a few thoughts:

This PR implements a new[1] downgrade protection mechanism introduced in TLS 1.3 for TLS 1.3 clients and servers. Will call this new mechanism `8446 Downgrade Protection(8446DP)` from here on. I will assume that when the spec says `TLS 1.3 server` it means a server endpoint configured to negotiate up to TLS 1.3. 

This text has me conflicted: `If negotiating TLS 1.1 or below...TLS 1.2 servers SHOULD, set the last 8 bytes of their ServerHello.Random`. A few implications:
- Since the downgrade value is embedded in the `random` field of the `ClientHello`/`ServerHello`, there's nothing technically preventing implementations from adding `8446DP` to TLS 1.2 clients and servers. 
- The spec(for TLS 1.3) is actually *encouraging* implementations to add this for "TLS 1.2" clients and servers  with a `SHOULD`. It would have been nice if the authors drafted this 1.2 recommendation in a separate RFC.

Brainstormed a few benefits/risks of implementing `8446DP` for s2n TLS 1.2 server. 

Cons(risks):
-  *every* TLS 1.2 handshake that negotiates with a < TLS 1.2 client will lose 8 bytes of randomness. Since TLS will still have 56 random bytes per handshake(32 client.random + 32 server.random - 8) this is likely a non issue. However current workloads with clients that have a low quality RNG may rely on the server endpoint to bring a bulk of the random bytes(at worst brings us down to ~24 bytes of randomness?).
- A buggy client that is configured to only use TLS 1.1 but has TLS 1.3 code paths may have a broken check for the `8446DP` magic bytes that causes handshakes to fail when we add `8446DP` to s2n.
- A buggy middlebox/firewall/anything that is misconfigured may choke on a TLS 1.2 ServerHello with the `8446DP` bytes set.

Pros(benefits):
- s2n's TLS 1.2 `8446DP` implementation will interop with other TLS libraries that implement `8446` DP for TLS 1.2 and provide a "better" downgrade protection mechanism than[1].

I am leaning slightly towards disabling `8446DP` for s2n TLS 1.2 server. Let me know if I missed anything or made a mistake.

[1] See https://tools.ietf.org/html/rfc7507 for how this is done for older versions of TLS with mixed effectiveness.

Reply:
I think openssl disables `8446DP` for TLS 1.2 per https://github.com/openssl/openssl/blob/bddbfae1cd667f0c1458deff98cbc03171e90d5a/ssl/statem/statem_lib.c#L1768-L1780 and https://github.com/openssl/openssl/blob/bddbfae1cd667f0c1458deff98cbc03171e90d5a/ssl/statem/statem_lib.c#L1846 

could also confirm by spinning up `s_client` and `s_server` compiled with 1.1.1